### DISCUSSION TITLE
How can I pass compile definitions to user C modules in cmake-based ports ? (couldn't succeed)

### DISCUSSION THREAD

**Discussion by scaprile** [micropython/micropython#13538] [General]
I have a bunch of C modules that I would like to use from micropython. Each of these modules has configuration options usually passed with -D on the compiler.
I'm not fond of cmake and not versed in it, I tried with what I could grasp:
```
add_definitions(-DTST1)
target_compile_options(usermod_test INTERFACE 
    -DTST2
)
target_compile_definitions(usermod_test INTERFACE 
    TST3
)
```
But no luck. My modules are built with a bunch of other files and the only definitions I see on the build command are those from the port -common and the extmod .cmake files. Not mine...

I attach a simple example with 2 Makefiles, one for the ESP32 and other for the RP2 ports. Both build OK if I remove the check for those definitions presence... and of course I don't see any of them in the generated build command.
[test.zip](https://github.com/micropython/micropython/files/14076364/test.zip)
The ESP32 example uses Docker to get the IDF image with everything installed. The RP2 example requires git and cmake. Both run on Linux.

I'd like to keep build flags inside user modules so they build with any port... of course.

**Comment by jimmo** [comment]
@scaprile of the three things you've tried, I think the third is correct:

```
target_compile_definitions(usermod_test INTERFACE 
    TST3
)
```

I tested this by adding

```c
#if FOO_BAR == 1
#error FOO_BAR==1
#endif
```

to the top of examples/usercmodule/cexample/examplemodule.c

Compiled with

```
$ make USER_C_MODULES=../../examples/usercmodule/cexample/micropython.cmake
```

It does not print the `#error` (as expected)

Then I added

```cmake
target_compile_definitions(usermod_cexample INTERFACE
    FOO_BAR=1
)
```

to the `micropython.cmake`, and recompiled, now I get

```
/home/jimmo/src/github.com/micropython/micropython/examples/usercmodule/cexample/examplemodule.c:8:2: error: #error FOO_BAR==1
    8 | #error FOO_BAR==1
      |  ^~~~~
```

**Reply by scaprile** [reply]
Hello @jimmo , thank you for answering.
Please let me point out that you are doing sort of an "anti-test", not what I tested.
The code I attached, like my actual code, tests for `defined`, not for value.
```
#if !defined(TST1) && !defined(TST2) && !defined(TST3)
```
Please notice that I also test for all three ways failing, so your way of testing should have passed my tests, and it didn't.
Also, from the number of dots in your command line, I guess you are testing the RP2 port, while I'm battling the IDF. Nevertheless, I went down a bit deeper and managed to get what is going on, and it happens in both ports.

I changed test code to something more debug friendly, just checking for pressence and later for value:
```
#ifndef myTST1
#warning myTST1 failed
#endif
#if myTST1 == 1
#warning myTST1 is 1
#endif
(same for 2 and 3, you get the picture...)
```

- First, there seems to be  some parsing for generation of some headers, that calls my source code with no definitions at all, and that triggers the error:
```
[  1%] Generating genhdr/qstr.i.last
/home/scaprile/micropython/examples/usercmodule/cexample/examplemodule.c:6:2: warning: #warning myTST1 failed [-Wcpp]
    6 | #warning myTST1 failed
      |  ^~~~~~~
/home/scaprile/micropython/examples/usercmodule/cexample/examplemodule.c:9:2: warning: #warning myTST2 failed [-Wcpp]
    9 | #warning myTST2 failed
      |  ^~~~~~~
/home/scaprile/micropython/examples/usercmodule/cexample/examplemodule.c:12:2: warning: #warning myTST3 failed [-Wcpp]
   12 | #warning myTST3 failed
      |  ^~~~~~~
```

- Then, there is the actual building of the code, where definitions and compile options are properly passed (though the behavior of `add_definitions()` is a bit weird and I would avoid it):
```
[ 74%] Building C object CMakeFiles/firmware.dir/home/scaprile/micropython/examples/usercmodule/cexample/examplemodule.c.obj
/home/scaprile/micropython/examples/usercmodule/cexample/examplemodule.c:15:2: error: #warning myTST1 is 1 [-Werror=cpp]
   15 | #warning myTST1 is 1
      |  ^~~~~~~
/home/scaprile/micropython/examples/usercmodule/cexample/examplemodule.c:18:2: error: #warning myTST2 is 2 [-Werror=cpp]
   18 | #warning myTST2 is 2
      |  ^~~~~~~
/home/scaprile/micropython/examples/usercmodule/cexample/examplemodule.c:21:2: error: #warning myTST3 is 3 [-Werror=cpp]
   21 | #warning myTST3 is 3
      |  ^~~~~~~
```

So far, the only way I found is to workaround this issue by adding my definitions to `ports/esp32/esp32_common.cmake`. As I pull 1.22.1 on clean, I `sed -i -e '/target_compile_definitions.*$$/a \ \ \ \ $(BUILD_DEFS)' -e '/target_compile_options.*$$/a \ \ \ \ $(BUILD_OPTS)' micropython/ports/esp32/esp32_common.cmake` inside the Makefile.

Is there a way to generate `genhdr/qstr.i.last` passing proper definitions ? Or at least to instruct it to just parse the micropython module specific code and not all the module files? My micropython module has one file for the module itself, wrapping the actual C module (which needs those definitions) that I've been using for decades.

Thanks for your help !

Addendum: I feel this triggers another possible problem. If the first pass, generation phase, which I guess translates those macros to micropython structures, runs with no defines on a module using conditional compilation to embed functions or not, it would fail. All Python bindings should be provided no matter what, and "not included" stubs added in place of those functions that have not been included in the C module built on the second pass...
